### ISSUE TITLE
slow memory retrieval (significantly slower then simple pickle)

### ISSUE THREAD

**Issue by MInner** [joblib/joblib#467]
Hi,

I'm little confused by why does reading and writing to (file based) "memory" take so enormous amount of time compared to bare pickling/unpickling.

In my case, `func()` is a tiny memorized function that takes a short string argument and returns a (short) dict with (long) lists of ~complex objects. For some reason, function retrieval from cache takes significantly more time then just unpickleing file. Resulting file is approximately 70Mb.

I observe same thing for any other function.

```
%prun func(some_str)

        1   12.436   12.436   52.011   52.011 pickle.py:1014(load)
 41531482    7.665    0.000   11.931    0.000 pickle.py:226(read)
  1922386    5.547    0.000    7.339    0.000 pickle.py:1504(load_build)
 41531483    4.266    0.000    4.266    0.000 {method 'read' of '_io.BufferedReader' objects}
  6490284    3.753    0.000    6.666    0.000 pickle.py:1439(load_long_binput)
  2645763    2.666    0.000    4.764    0.000 pickle.py:1192(load_binunicode)
 30070039    2.403    0.000    2.403    0.000 {built-in method builtins.isinstance}
  4140172    1.870    0.000    3.225    0.000 pickle.py:1415(load_binget)
  1922386    1.369    0.000    2.049    0.000 pickle.py:1316(load_newobj)
  9196954    1.359    0.000    1.359    0.000 {built-in method _struct.unpack}
  1922386    1.114    0.000    8.724    0.000 numpy_pickle.py:319(load_build)
 10857316    0.962    0.000    0.962    0.000 {method 'pop' of 'list' objects}
 14536246    0.873    0.000    0.873    0.000 {method 'append' of 'list' objects}
  1922386    0.873    0.000    1.218    0.000 pickle.py:1472(load_setitem)
  1922393    0.816    0.000    0.816    0.000 {built-in method builtins.getattr}
   676815    0.765    0.000    1.384    0.000 pickle.py:1458(load_appends)
  1922387    0.730    0.000    0.832    0.000 pickle.py:1257(load_empty_dictionary)
        1    0.715    0.715   53.099   53.099 <string>:1(<module>)
  1245385    0.559    0.000    0.848    0.000 pickle.py:1451(load_append)
...

%prun len(pickle.load(open("..file..", 'rb')))
        1    4.587    4.587    4.587    4.587 {built-in method _pickle.load}
        1    0.553    0.553    5.140    5.140 <string>:1(<module>)
        1    0.000    0.000    5.140    5.140 {built-in method builtins.exec}
        1    0.000    0.000    0.000    0.000 {built-in method io.open}
        1    0.000    0.000    0.000    0.000 {built-in method builtins.len}
        1    0.000    0.000    0.000    0.000 {method 'disable' of '_lsprof.Profiler' objects}
```

**Comment by lesteve** [comment]
From the profiles I would guess this is because joblib is using the python pickle, mostly to be able to hack into the pickle mechanism and specialise it for numpy arrays. You are probably using python 3 for which pickle is the C pickle. Have a look at  #421 for example where this was discussed in more details.

**Comment by ogrisel** [comment]
Indeed. joblib is not very useful when dealing with Python objects with a large number of small sub objects. It's more useful for Python objects with a very few large subobjects (e.g. large numpy arrays).

I am afraid that there is little we can do about it.

**Comment by GaelVaroquaux** [comment]
> I am afraid that there is little we can do about it.

Would it be possible to write in the stored file whether or not there are
numpy objects in it, and if there are none use the standard unpickler at
load time?

**Comment by ogrisel** [comment]
Walking the graph structure of such complex objects is actually quite costly.

**Comment by GaelVaroquaux** [comment]
> Walking the graph structure of such complex objects is actually quite costly.

But once the storage is finished, it's done, so after storing, we should
know.

**Comment by lesteve** [comment]
In the Memory case, I guess whether output.pkl contains a numpy array or not could be written in a metadata file in the same folder (e.g. in metadata.json). This way you know at load-time you can use pickle.load directly rather than joblib.load?

**Comment by aabadie** [comment]
> But once the storage is finished, it's done, so after storing, we should know.

This only solves half of the problem : storing is still slow but loading is fast. Also storing in the file itself whether it contains or not a numpy array also introduces another format change with backward compatibility necessity.

**Comment by aabadie** [comment]
> In the Memory case, I guess whether output.pkl contains a numpy array or not could be written in a metadata file in the same folder (e.g. in metadata.json). This way you know at load-time you can use pickle.load directly rather than joblib.load?

Sounds like a potential solution. So `pickle.load` should return something indicating whether there's a numpy array or not ? And the memory API is responsible of using it or not.

**Comment by aabadie** [comment]
`s/pickle.load/pickle.dump/` in previous comment

**Comment by lesteve** [comment]
Actually thinking about it, maybe the cleanest thing to do is to add a `use_joblib_pickling` argument to `Memory`, which should be True by default.

**Comment by GaelVaroquaux** [comment]
> Actually thinking about it, maybe the cleanest thing to do is to add a
> use_joblib_pickling argument to Memory, which should be True by default.

That would work, but the solution of storing whether or not it is a
standard pickle would be automatic for the user, no?

I am not sure anymore: if there are no numpy arrays, is the stored file a
standard pickle or isn't it?

**Comment by aabadie** [comment]
> if there are no numpy arrays, is the stored file a standard pickle or isn't it?

it is

**Comment by GaelVaroquaux** [comment]
> it is

Awesome, so if we know that there are no numpy arrays in it, we can use
the fast loading route, right?

**Comment by aabadie** [comment]
> so if we know that there are no numpy arrays in it, we can use the fast loading route, right?

Yes but it can be costly to determine whether the arbitrary object contains a numpy array or not (if we use the python implementation of pickle).

**Comment by GaelVaroquaux** [comment]
> Yes but it can be costly to determine whether the arbitrary object contains a
> numpy array or not (if we use the python implementation of pickle).

At write time we can insert a code path in our pickler which detects
that.

**Comment by aabadie** [comment]
> a code path 

What do you mean ?

**Comment by GaelVaroquaux** [comment]
> What do you mean ?

In our NumpyPickler, we can change a flag (on the pickler for instance)
when we hit an array.

**Comment by aabadie** [comment]
> In our NumpyPickler, we can change a flag (on the pickler for instance) when we hit an array.

So if I understand correctly, at the end of the dump, this flag will be written at the beginning of the pickle file ? This will change the pickle format once again and we'll have to deal with 3 pickle formats:
* valid pickle with no numpy arrays
* old invalid pickles with numpy arrays, introduced in 0.10
* new invalid pickles with numpy arrays
That will be a pleasure to maintain ;)

**Comment by GaelVaroquaux** [comment]
> So if I understand correctly, at the end of the dump, this flag will be written
> at the beginning of the pickle file ?

I think that we could save it in the metadata of memory object.

**Comment by GaelVaroquaux** [comment]
> I think that we could save it in the metadata of memory object.

I mean of the memory entry. What is currently stored in the .json file.

It's a bit of a hack, granted.

**Comment by aabadie** [comment]
> I mean of the memory entry. What is currently stored in the .json file.

Ok

At load time, what about a try/except strategy ? :
```
try:
    pickle.load()
except:
   # Can fail if there's a numpy array in the pickle.
   joblib.load()
```
If there's no numpy array the load will use the C implementation of pickle otherwise, it will fail and use the joblib fallback mecanism.

**Comment by lesteve** [comment]
> At load time, what about a try/except strategy ? :

I thought about that. The problem is the array may be at the very end of the pickle so you waste a lot of time almost all the object and ultimately failing before trying the `joblib.load` strategy.

I don't think we can change the pickle format (adding the info somewhere in the file whether an array is present in the pickle) while still:
* keeping streamability, which was one of the feature we strived to have during the single-file pickle PR
* avoiding to walk the entire object to check to figure out whether an array is present or not

> I think that we could save it in the metadata of memory object.

We could do that (I mentioned this option in https://github.com/joblib/joblib/issues/467#issuecomment-270848836).

I find adding a `use_joblib_pickling` argument a bit cleaner (as mentioned in https://github.com/joblib/joblib/issues/467#issuecomment-271266252), for example there is one less file to read at `joblib.load` time (metadata.json is not read when we do `joblib.load` currently). I guess the inconvenient with the `use_joblib_pickling` is the user has to make an informed decision.

**Comment by rth** [comment]
Also, in the context of scikit-learn, the assumption is that estimators should be saved  / loaded with `joblib`. However, for instance a `TfidfVectorizer` object (which contains no numpy arrays) is ~ 50x slower to load with joblib than with Python 3 pickle..

**Comment by ssfrr** [comment]
just to add another use-case for this, I'm trying out using joblib to cache sets of compiled Theano functions so I don't have to be constantly re-building them when I'm developing and re-starting things frequently. I previously rolled my own hash/pickle caching, but using joblib is much more elegant. The function sets end up being 200-250MB uncompressed, and loading/storing them is about 50-75% slower than when using pickle directly (this is on python 3.6, where pickle == cPickle).

**Comment by ssfrr** [comment]
whoops - correction: it seems that when I turn compression off the speeds are actually pretty comparable, but the size on disk seems about twice as big as when I used pickle directly (using HIGHEST_PROTOCOL). Sorry for the mis-leading info.

**Comment by GaelVaroquaux** [comment]
Did you try using lzma? I seem to remember that it's a pretty fast
compression algorithm.

**Comment by aabadie** [comment]
> I seem to remember that it's a pretty fast compression algorithm.

It depends on the data but generally it compress better with the cost of slower speed.

**Comment by GaelVaroquaux** [comment]
> It depends on the data but generally it compress better with the cost of slower speed.

I've already made this mistake: I keep confusing it with another one
(lz4, I believe). Maybe it would be useful to add in the joblib doc hints
about which is fast and which compresses better.

**Comment by 22459** [comment]
So, is there any hope for getting memory.cache to use normal pickle? It seems to slow down my load time by an order of magnitude.

**Comment by lesteve** [comment]
Here is a work-around. It may be a bit brittle, meaning that it might break on the next joblib release, but maybe that helps in your case.

```py
import pickle
import os

import joblib


class MyMemory(joblib.Memory):
    def _load_output(output_dir, func_name, timestamp=None, metadata=None,
                     mmap_mode=None, verbose=0):
        """Load output of a computation."""
        if verbose > 1:
            signature = ""
            try:
                if metadata is not None:
                    args = ", ".join(['%s=%s' % (name, value)
                                      for name, value
                                      in metadata['input_args'].items()])
                    signature = "%s(%s)" % (os.path.basename(func_name),
                                                 args)
                else:
                    signature = os.path.basename(func_name)
            except KeyError:
                pass

            if timestamp is not None:
                t = "% 16s" % format_time(time.time() - timestamp)
            else:
                t = ""

            if verbose < 10:
                print('[Memory]%s: Loading %s...' % (t, str(signature)))
            else:
                print('[Memory]%s: Loading %s from %s' % (
                        t, str(signature), output_dir))

        filename = os.path.join(output_dir, 'output.pkl')
        if not os.path.isfile(filename):
            raise KeyError(
                "Non-existing cache value (may have been cleared).\n"
                "File %s does not exist" % filename)
        result = numpy_pickle.load(filename, mmap_mode=mmap_mode)

    # May not need to override _persist_output if your write time are not a
    # problem
    def _persist_output(self, output, dir):
        """ Persist the given output tuple in the directory.
        """
        try:
            filename = os.path.join(dir, 'output.pkl')
            mkdirp(dir)
            write_func = pickle.dump
            concurrency_safe_write(output, filename, write_func)
            if self._verbose > 10:
                print('Persisting in %s' % dir)
        except OSError:
            " Race condition in the creation of the directory "


mem = MyMemory('/tmp/test')

def identity(x):
    return x

cached_identity = mem.cache(identity)
cached_identity(3)
cached_identity(3)
```

Maybe we could make it easier to override the `dump` and `load` function in derived classes of Memory. Not entirely sure whether there is hope of doing it in a cleaner way inside joblib. See https://github.com/joblib/joblib/issues/467#issuecomment-271345741 for what I think is a decent summary. You may be able to find more context if you search through issues.

**Comment by 22459** [comment]
Thanks, I'll give it a shot.

**Comment by GaelVaroquaux** [comment]
I think that there is some hope that these problems can be tackled in
future version of Python, as the C pickler is getting better and more
versatile.

**Comment by 22459** [comment]
I think I got the following patch to memory.py to work:
```diff --git a/joblib/joblib/memory.py b/joblib/joblib/memory.py
index 14d7552..536826e 100644
--- a/../joblib/joblib/memory.py
+++ b/joblib/memory.py
@@ -34,7 +34,7 @@ from .func_inspect import format_call
 from .func_inspect import format_signature
 from ._memory_helpers import open_py_source
 from .logger import Logger, format_time, pformat
-from . import numpy_pickle
+import pickle
 from .disk import mkdirp, rm_subdirs, memstr_to_bytes
 from ._compat import _basestring, PY3_OR_LATER
 from .backports import concurrency_safe_rename
@@ -134,7 +134,7 @@ def _load_output(output_dir, func_name, timestamp=None, metadata=None,
         raise KeyError(
             "Non-existing cache value (may have been cleared).\n"
             "File %s does not exist" % filename)
-    result = numpy_pickle.load(filename, mmap_mode=mmap_mode)
+    result = pickle.load(filename)
 
     return result
 
@@ -208,7 +208,7 @@ def concurrency_safe_write(to_write, filename, write_func):
     thread_id = id(threading.current_thread())
     temporary_filename = '{}.thread-{}-pid-{}'.format(
         filename, thread_id, os.getpid())
-    write_func(to_write, temporary_filename)
+    write_func(to_write, open(temporary_filename,"wb"))
     concurrency_safe_rename(temporary_filename, filename)
 
 
@@ -759,8 +759,7 @@ class MemorizedFunc(Logger):
         try:
             filename = os.path.join(dir, 'output.pkl')
             mkdirp(dir)
-            write_func = functools.partial(numpy_pickle.dump,
-                                           compress=self.compress)
+            write_func = pickle.dump
             concurrency_safe_write(output, filename, write_func)
             if self._verbose > 10:
                 print('Persisting in %s' % dir)
```
Of course its a huge hack that just bypasses everything.

**Comment by 22459** [comment]
Idk what joblib does, but wrapping a function that takes 20 seconds to run makes it take 70 seconds to run...

**Comment by lesteve** [comment]
That's does not seem ideal indeed. Have you made sure that you were using the C pickle module? It is named pickle on Python 3 but cPickle under Python 2. My 2c: the best way to get good feed-back is to give a snippet that reproduces your problem, if you could do that it would be great. Maybe in your use case, using joblib.Memory is not a good idea, hard to tell without seeing some code.

Joblib stores the result of the function in a pickle (with some optimization for numpy arrays) the first time around. The second time around it detects that the function has been called and loads the result from disk. If you think the [doc](http://joblib.readthedocs.io/en/latest/memory.html) can be improved to make it more clear, you are more than welcome to open a PR!

If you believe that joblib could do better performance-wise, you are more than welcome to look at the source code and open a PR too.  As Gael was saying there may be improvements in the stdlib pickle for Python 3.7 (or maybe Python 3.8, not sure) that makes some of the joblib hacks unnecessary.

**Comment by 22459** [comment]
Yeah, that's more or less what I figured. I'm still unsure where the issue is stemming from, and at the moment my code is such a mess I cant get it to work anyway. I'll see later.

**Comment by gingsi** [comment]
In case anyone also stumbles upon this still open issue.

With large nested dictionaries, the difference can be huge: I currently get numpy_pickle 70.3s vs pickle 1.6s. (44x slower) for a ~50MB pickle file.

If you don't need the numpy pickling functionality, use it like this:

```
# file define_backend.py
import os
import os.path
import pickle

from joblib import register_store_backend
from joblib._store_backends import FileSystemStoreBackend


class FileSystemStoreBackendNoNumpy(FileSystemStoreBackend):
    NAME = "no_numpy"

    def load_item(self, path, verbose=1, msg=None):
        """Load an item from the store given its path as a list of
           strings."""
        full_path = os.path.join(self.location, *path)

        if verbose > 1:
            if verbose < 10:
                print('{0}...'.format(msg))
            else:
                print('{0} from {1}'.format(msg, full_path))

        mmap_mode = (None if not hasattr(self, 'mmap_mode')
                     else self.mmap_mode)

        filename = os.path.join(full_path, 'output.pkl')
        if not self._item_exists(filename):
            raise KeyError("Non-existing item (may have been "
                           "cleared).\nFile %s does not exist" % filename)

        # file-like object cannot be used when mmap_mode is set
        if mmap_mode is None:
            with self._open_item(filename, "rb") as f:
                item = pickle.load(f)
        else:
            item = pickle.load(filename, mmap_mode=mmap_mode)
        return item


register_store_backend(FileSystemStoreBackendNoNumpy.NAME, FileSystemStoreBackendNoNumpy)
```

```
# file use_memory.py
from define_backend import FileSystemStoreBackendNoNumpy
mem = joblib.Memory(backend=FileSystemStoreBackendNoNumpy.NAME)

@mem.cache
def function_to_cache():
    pass  # etc
```
